src/Features/Core/Portable/AddImport/AbstractAddImportFeatureService.cs

```diff
@@ -475,5 +470,102 @@ private static bool NotNull(SymbolReference reference)
+            var symbolSearchService = searchReferenceAssemblies || searchNuGetPackages
+                ? solution.Workspace.Services.GetService<ISymbolSearchService>()
+                : null;
+
+            var packageSources = symbolSearchService != null && searchNuGetPackages
+                ? GetPackageSources(document)
+                : ImmutableArray<PackageSource>.Empty;
+
+            return GetFixesForDiagnosticsAsync(document, span, diagnostics, symbolSearchService, searchNuGetPackages, packageSources, cancellationToken);
+        }
+
+        public async Task<ImmutableArray<(Diagnostic Diagnostic, ImmutableArray<AddImportFixData> Fixes)>> GetFixesForDiagnosticsAsync(
+            Document document, Text.TextSpan span, ImmutableArray<Diagnostic> diagnostics,
+            ISymbolSearchService symbolSearchService, bool searchReferenceAssemblies, 
+            ImmutableArray<PackageSource> packageSources, CancellationToken cancellationToken)
+        {
+            var addImportService = document.GetLanguageService<IAddImportFeatureService>();
+
+            // We might have multiple different diagnostics covering the same span.  Have to
+            // process them all as we might produce different fixes for each diagnostic.
+
+            var documentOptions = await document.GetOptionsAsync(cancellationToken).ConfigureAwait(false);
+            var placeSystemNamespaceFirst = documentOptions.GetOption(GenerationOptions.PlaceSystemNamespaceFirst);
+
+            var fixesForDiagnosticBuilder = ArrayBuilder<(Diagnostic, ImmutableArray<AddImportFixData>)>.GetInstance();
+
+            foreach (var diagnostic in diagnostics)
+            {
+                var fixes = await addImportService.GetFixesAsync(
+                    document, span, diagnostic.Id, placeSystemNamespaceFirst,
+                    symbolSearchService, searchReferenceAssemblies,
+                    packageSources, cancellationToken).ConfigureAwait(false);
+
+                fixesForDiagnosticBuilder.Add((diagnostic, fixes));
+            }
+
+            return fixesForDiagnosticBuilder.ToImmutableAndFree();
+        }
+
+        public ImmutableArray<CodeAction> GetCodeActionsForFixes(Document document, ImmutableArray<AddImportFixData> fixes, IPackageInstallerService installerService = null)
+        {
+            var codeActionsBuilder = ArrayBuilder<CodeAction>.GetInstance();
+
+            foreach (var fix in fixes)
+            {
+                var codeAction = TryCreateCodeAction(document, fix, installerService);
+
+                codeActionsBuilder.AddIfNotNull(codeAction);
+
+                if (codeActionsBuilder.Count >= MaxResults)
+                {
+                    break;
+                }
+            }
+
+            return codeActionsBuilder.ToImmutableAndFree();
+        }
+
+        public CodeAction TryCreateCodeAction(Document document, AddImportFixData fixData, IPackageInstallerService installerService = null)
```

Comment:
when is this passed in as null?  can we make this non optional to make it clear?

Reply:
I was trying to preserve the ability to pass mocks in for testing. This will only be non-null in tests.